### ISSUE TITLE
onBecomeObserved won't trigger, in case of observable class changed

### ISSUE THREAD

**Issue by kuzmichDev** [mobxjs/mobx#2667]
Hello,
When i tried to change instance inside some kind of store, with onBecomeObserved, it wont triggers

There is example: https://codesandbox.io/s/onbecomeobserved-wont-trigger-86q8d

**Comment by urugator** [comment]
It's caused by that `@computed` on `store.items` possibly related to https://github.com/mobxjs/mobx/issues/2309#issuecomment-678376447, not sure.
CC @mweststrate

**Comment by mweststrate** [comment]
@kuzmichDev please don't store state in `useMemo`, memo doesn't guarantee to remember your state, so you'll need useState instead. Beyond that, can you describe what you are trying to achieve? I'm not sure if listening to onBecomeObserved on a field means what you think it means. The behavior of your current approach will be very confusing regardless how it works, as you are changing the same thing as an effect of observing a change

**Comment by kuzmichDev** [comment]
@mweststrate
Hi, i'm trying to achieve something like this:

`LazyInitializedList` empty, if nobody observes `items` inside.
If some component start observing `items` (initialized as `undefined` by default), `LazyInitializedList` will fetch some data and then set it to observing field `items` (in my example it just set items from variable).
When observer of that items disappeared (unmounted etc.), `LazyInitializedList` will clear that field.

Also, i need to change instances of that `LazyInitializedList` (in case of configuration change, etc).

So i'm expect, that `items` from new `LazyInitializedList` would becomes observable also (it should through computed) and new instance of `LazyInitializedList` will do fetch and set them in observed variable 'items', but if previous list (initial) already was observed  `onBecomeObserved` wont triggers for new list.

in console, i expect to see:
```
onBecomeObserved initial
// button press
onBecomeUnobserved initial
onBecomeObserved new // wont happen
```

**Comment by mweststrate** [comment]
Sorry for the late follow up, stalebot doesn't respect holidays :-P. 

In MobX derivations should not alter state. Events like onBecomeObserved and onBecomeUnobserved as part of the derivation lifecycle, so making changes directly in them might not work as expected, because you are changing the underlying data is swapped _during_ the computation, so handling cases like these will always result in unpredictable behavior (like unexpected double derivations etc).  Those events are not meant to update observables, but rather to start or suspend other processes, such as a web socket example. 

Instead, the proper tool to build utilities like this is usually using `createAtom`, see: https://mobx.js.org/custom-observables.html, where you can initialize data upon the first read from your data structure. But for a minimal change that makes your example work, is to make sure that the data is initialized _before_ it is read for the first time, rather than after: https://codesandbox.io/s/onbecomeobserved-wont-trigger-forked-29ku0?file=/src/App.tsx

**Comment by mweststrate** [comment]
Should be fixed in 6.0.5
